### DISCUSSION TITLE
Usage as an information extraction tool for CSS

### DISCUSSION THREAD

**Discussion by kraklin** [jfmengels/elm-review#48] [Ideas]
Hey Jeroen,

I have been reading your blogpost and the `Usage as an information extraction tool` paragraph caught my attention. Particularly this part: 

> For instance if you use CSS files and classes, you could extract the list of CSS classes currently being used in your application, and pass that to your CSS linter which would tell you to remove the unused classes.

which is exactly the idea I was having for a long time and I have ran into the need for this yesterday as well. 

We are using [Weak-CSS methodology](https://github.com/GlobalWebIndex/weak-css) and [class-namespaces](https://github.com/GlobalWebIndex/class-namespaces) to build our styles, which makes any tool like `post-css` useless :( The only solution would be to somehow parse the `class-namespaces` calls and reconstruct the CSS class name out of that and then compare it with the class names from SASS to find out which classes are not used at all.

Another nice use-case for that would be using `tailwind-css` where the resulted CSS file can be stripped down only to used classes by PostCSS, but only if you use directly the whole `tailwind` classname. You can't concat those names or do any other optimisations on those as well, which is something I'd definitely use :)

So my questions are:
  - can elm-review help in those situations?
  - how should it work with elm-review to have those custom use-cases (I bet there would be many different use-cases on how to put class names together)
  - what is currently the biggest problem to move towards making the information extraction from elm-review?

**Comment by jfmengels** [comment]
> can elm-review help in those situations?

I feel like you could definitely do this yes.

> The only solution would be to somehow parse the class-namespaces calls and reconstruct the CSS class name out of that and then compare it with the class names from SASS to find out which classes are not used at all.

From what I read in the [usage documentation](https://github.com/GlobalWebIndex/class-namespaces#usage), you want to track all the `Weakspace.namespace` and `WeakCss.add` values so that you can then reconstruct them manually when `WeakCss.toClass` and similar are used. It's not something I have had to do yet, but I don't think it will be all that complex (well, maybe if it's your first `elm-review` rule).

There will be cases where you can't know the result without very extensive flow analysis, for instance if you have a function that allows for dynamic parts.

```elm
view string className =
  div [WeakCss.add string className |> WeakCss.toClass] []
```

What you can then do is report an error, saying something like "the code is too smart for me, please simplify it so that I can make sure everything works as expected".

I have an alternative solution that you could use already instead of using this an extraction rule, but which would be easy to transform to that later on, so it's worth trying out already.

Instead of having `elm-review` extract data out of the codebases, you could from your CSS files generate an Elm file with the list of CSS classes available, then have your `elm-review` rule (describe above) import that list and use it as configuration.

The benefits of that is you will in your Elm code be able to tell "hey, no CSS rule exists for this class. Please check if you haven't made a typo, or create the CSS rule". That will give you more actionable feedback when working with your Elm code. Then you can mark the classes are used, and report the ones that have not been used as a "global" error (there are no global errors yet, but basically you can create an error that just points somewhere). If the generated code contains the filename and line numbers, you can potentially give some good indications as to where in the CSS files you need to go to delete the code.

I like this solution maybe more than the one with extracting data. The downside is that you will need to run this generator before every `elm-review` invocation, and potentially have a watcher that re-generates the file if you want to use `elm-review` in watch mode.

> what is currently the biggest problem to move towards making the information extraction from elm-review?

I recently basically implemented the biggest hurdle to make the [module name lookup table](https://package.elm-lang.org/packages/jfmengels/elm-review/latest/Review-Rule#withModuleNameLookupTable) work. Basically, before running your rules, I run this native rule, extract the data it collected and pass that information to the other rules for them to consume.

The rest is mostly design issues around the user experience and the API.

I have described them in more detail in jfmengels/elm-review-design-discussions#7.

**Comment by kraklin** [comment]
> Instead of having elm-review extract data out of the codebases, you could from your CSS files generate an Elm file with the list of CSS classes available, then have your elm-review rule (describe above) import that list and use it as configuration.

So you propose to create Elm file with list of all generated class names, e.g. 

```css
// CSS
.namespace__module--other{... }
.namespace__module{... }

-- Elm
classes = 
  [ 'namespace__module'
  , 'namespace__module--other'
  ]
```

and then use that `classes` List in elm-review rule which will crawl through all `WeakCss.namespace` and `WeakCss.add` etc. and remove that the name from the List. And if there will be anything in the list left, those are (potential) candidates to be removed. 

>  It's not something I have had to do yet, but I don't think it will be all that complex (well, maybe if it's your first elm-review rule).

Yes, it will be my first rule and I have no idea how to do it yet :D

**Comment by jfmengels** [comment]
Yes exactly. A bit more details here:

- Everytime you encounter usage of an unknown class, create an error at the call site of `WeakCss.toClass` and friends
- You will need to make a "project rule" so you can traverse the whole project and keep in "context" what CSS classes were used
- Rather than removing those names from the list, you should have a separate "used" list (a `Set` probably actually), and at the end of the project review using [`withFinalProjectEvaluation`](https://package.elm-lang.org/packages/jfmengels/elm-review/latest/Review-Rule#withFinalProjectEvaluation) you'll diff the two and report the ones that were never used.
